### ISSUE TITLE
Retrying of failed downloads

### ISSUE THREAD

**Issue by marcbull** [microsoft/vcpkg#42478]
Package builds randomly fail for packages downloaded from gitlab.gnome.org. After a few retries, the build magically succeeds and fails with another package from the same domain.

```
[DEBUG] 1000: execute_process(curl --fail -L https://gitlab.gnome.org//GNOME/libadwaita/-/archive/1.3.2/libadwaita-1.3.2.tar.gz --create-dirs --output /workspaces/mxl-player/vcpkg/downloads/GNOME-libadwaita-1.3.2.tar.gz.4115.part)
[DEBUG] 1000: cmd_execute_and_stream_data() returned 22 after    58949 us
error: Missing GNOME-libadwaita-1.3.2.tar.gz and downloads are blocked by x-block-origin.
error: https://gitlab.gnome.org//GNOME/libadwaita/-/archive/1.3.2/libadwaita-1.3.2.tar.gz: curl failed to download with exit code 22
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed

curl: (22) The requested URL returned error: 503
```
Curl provides several options for retrying failed downloads: https://everything.curl.dev/usingcurl/downloads/retry.html 

**Environment
- OS: Linux/macOS

I was able to reproduce the behavior by running the curl command in a terminal. By adding the `--retry 4` parameter, the download was successful after 3 retries.

**Comment by FrankXie05** [comment]
@marcbull This situation may be caused by network fluctuations or server load. You can stop the vcpkg installation and try again.
```
Installing 8/9 gtk:x64-windows@4.16.3...
Building gtk:x64-windows@4.16.3...
-- Found Python version '3.12.7 at G:/bud/vcpkg/downloads/tools/python/python-3.12.7-x64/python.exe'
-- Using meson: G:/bud/vcpkg/downloads/tools/meson-1.6.0-ac58a9/meson.py
error: Missing GNOME-gtk-4.16.3.tar.gz and downloads are blocked by x-block-origin.
error: https://gitlab.gnome.org//GNOME/gtk/-/archive/4.16.3/gtk-4.16.3.tar.gz: failed: status code 503
CMake Error at scripts/cmake/vcpkg_download_distfile.cmake:231 (message):
  Download failed, halting portfile.
Call Stack (most recent call first):
  scripts/cmake/vcpkg_from_gitlab.cmake:113 (vcpkg_download_distfile)
  ports/gtk/portfile.cmake:4 (vcpkg_from_gitlab)
  scripts/ports.cmake:196 (include)


error: building gtk:x64-windows failed with: BUILD_FAILED
See https://learn.microsoft.com/vcpkg/troubleshoot/build-failures?WT.mc_id=vcpkg_inproduct_cli for more information.
Elapsed time to handle gtk:x64-windows: 4.1 s
Please ensure you're using the latest port files with `git pull` and `vcpkg update`.
Then check for known issues at:
  https://github.com/microsoft/vcpkg/issues?q=is%3Aissue+is%3Aopen+in%3Atitle+gtk
You can submit a new issue at:
  https://github.com/microsoft/vcpkg/issues/new?title=[gtk]+Build+error+on+x64-windows&body=Copy+issue+body+from+G%3A%2Fbud%2Fvcpkg%2Finstalled%2Fvcpkg%2Fissue_body.md
You can also submit an issue by running (GitHub CLI must be installed):
  gh issue create -R microsoft/vcpkg --title "[gtk] Build failure on x64-windows" --body-file G:/bud/vcpkg/installed/vcpkg/issue_body.md
PS G:\bud\vcpkg> ./vcpkg install libadwaita --recurse
Computing installation plan...
The following packages will be built and installed:
  * gtk:x64-windows@4.16.3
    libadwaita:x64-windows@1.3.2#2
Additional packages (*) will be modified to complete this operation.
Detecting compiler hash for triplet x64-windows...
Compiler found: C:/Program Files/Microsoft Visual Studio/2022/Community/VC/Tools/MSVC/14.39.33519/bin/Hostx64/x64/cl.exe
Installing 1/2 gtk:x64-windows@4.16.3...
Building gtk:x64-windows@4.16.3...
-- Found Python version '3.12.7 at G:/bud/vcpkg/downloads/tools/python/python-3.12.7-x64/python.exe'
-- Using meson: G:/bud/vcpkg/downloads/tools/meson-1.6.0-ac58a9/meson.py
Downloading https://gitlab.gnome.org//GNOME/gtk/-/archive/4.16.3/gtk-4.16.3.tar.gz
-- Extracting source G:/bud/vcpkg/downloads/GNOME-gtk-4.16.3.tar.gz
-- Applying patch 0001-build.patch
-- Applying patch fix_vulkan_enabled.patch
-- Using source at G:/bud/vcpkg/buildtrees/gtk/src/4.16.3-43597d369f.clean
-- Using cached msys2-mingw-w64-x86_64-pkgconf-1~2.3.0-1-any.pkg.tar.zst.
-- Using cached msys2-msys2-runtime-3.5.4-2-x86_64.pkg.tar.zst.
-- Using msys root at G:/bud/vcpkg/downloads/tools/msys2/21caed2f81ec917b
-- Found external ninja('1.11.0').
-- Configuring x64-windows-dbg
-- Getting CMake variables for x64-windows
-- Configuring x64-windows-dbg done
-- Configuring x64-windows-rel
-- Configuring x64-windows-rel done
-- Package x64-windows-dbg
```

**Comment by marcbull** [comment]
@FrankXie05 Yes, error 503 can be caused by high traffic spikes or resource exhaustion. But unfortunately, it seems, the gitlab.gnome.org server is extremely limited in its resources. This leads to an extremely high and frustrating failure rate for CI builds, and this would be an easy workaround for this problem.

As an alternative, the GitHub mirrors for the gnome projects could be used.

**Comment by jeffro59** [comment]
I got around this frustrating issue by downloading all the files, and then ran the install command again to resume building where it failed.  Next time I'll download all before I attempt an install.  Below is the command to download for a package (boost in this example).  It will download just the files needed as the downloads are cached in a download folder from the failed build.

vcpkg install boost --only-downloads
